fix(sendconfig): Do not skip when SHA is not changed if the update is performed by a fallback

randmonkey opened the pull request:
<!-- Thanks for sending a pull request! Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#the-pull-request-submit-process and developer guide https://github.com/kubernetes/community/blob/master/contributors/devel/development.md
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://github.com/kubernetes/community/blob/master/contributors/guide/release-notes.md and ensure your changes are being reflected in CHANGELOG.md for the next upcoming release
-->

**What this PR does / why we need it**:

<!-- Please describe why this particular PR is necessary or why you see it as a nice addition -->

Do not skip the update of config if the SHA is the same as last valid configuration when the update is performed by a fallback. This fixes the issue that in DB mode, if the fallback configuration is the same as the last  valid configuration, the entities that should be purged together with the invalid entities still remains.

**Which issue this PR fixes**:

<!--
Here you can add any links to issues that this PR is relevant for.
You can use Github keywords (like: closes, fixes or resolves) to auto-resolve
the linked issue(s) when this PR gets merged.

For example: fixes #<issue number>
-->
fixes #7374 

**Special notes for your reviewer**:

<!-- Here you can add any open questions or notes that you might have for reviewers -->

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

pmalek commented on internal/dataplane/sendconfig/sendconfig.go:
```diff
@@ -58,8 +58,11 @@ func PerformUpdate(
+	// - reverse sync is enabled
+	// - or the update is performed by a fallback process because in DBmode, when applying fails, some entities are applied successfully, then the Kong gateway may be in a "partial success" state.
+	// REVIEW: Only disable the check on fallback in DBmode?
+	if !config.EnableReverseSync || !isFallback {
```
`!config.EnableReverseSync || !isFallback` != `!(config.EnableReverseSync || isFallback)`


| config.EnableReverseSync | isFallback | !config.EnableReverseSync OR !isFallback | !(config.EnableReverseSync OR isFallback) | !config.EnableReverseSync AND !isFallback |
|--------------------------|------------|------------------------------------------|-------------------------------------------|-------------------------------------------|
| 0                        | 0          | 1                                        | 1                                         | 1                                         |
| 1                        | 0          | 1                                        | 0                                         | 0                                         |
| 0                        | 1          | 1                                        | 0                                         | 0                                         |
| 1                        | 1          | 0                                        | 0                                         | 0                                         |

```suggestion
	if !(config.EnableReverseSync || isFallback) {
```

randmonkey replied:
Fixed.

pmalek commented on internal/dataplane/sendconfig/sendconfig.go:
```diff
@@ -58,8 +58,11 @@ func PerformUpdate(
 
-	// disable optimization if reverse sync is enabled
-	if !config.EnableReverseSync {
+	// Disable the optimization by checking if the configuration is same as the last applied configuration if:
```
```suggestion
	// Disable the optimization of checking if the configuration is the same as the last applied configuration if:
```

pmalek commented on internal/dataplane/sendconfig/sendconfig.go:
```diff
@@ -58,8 +58,11 @@ func PerformUpdate(
+	// Disable the optimization by checking if the configuration is same as the last applied configuration if:
+	// - reverse sync is enabled
+	// - or the update is performed by a fallback process because in DBmode, when applying fails, some entities are applied successfully, then the Kong gateway may be in a "partial success" state.
+	// REVIEW: Only disable the check on fallback in DBmode?
```
I _think_ that we can disable this check in both DB mode only. When working with dbless it's a simple hash to compare whether there's need apply new config.

So the condition below would need to change.

randmonkey commented:
The `TestConsumerGroup` integration test case fails after fixing this. From further investigations, I found that when the `response-transformer-advanced` plugin for route + consumer cannot be created with the following error:
```
performing update for http://172.18.128.2:8001 failed: 2 errors occurred:\n\twhile post processing event: inserting plugin response-transformer-advanced for route 7d58209a-68dd-5879-a029-08799f6c67fb and consumer-group 0785e91f-262c-51fd-bb7e-baa9b7b644d4: entity already exists\n\twhile post processing event: inserting plugin response-transformer-advanced for route 7d58209a-68dd-5879-a029-08799f6c67fb and consumer-group 393341ed-a05d-504b-b826-db76f9610dba: entity already exists\n"
```
But I can also see the event to create the same plugin in the logs:
```
updated gateway entity	{"v": 1, "action": "create", "kind": "plugin", "name": "response-transformer-advanced for route e60ad185-9a86-4ad1-a22f-97c3a93c5db4.multi.echo..1027 and consumer-group test-consumer-group-1"}
updated gateway entity	{"v": 1, "action": "create", "kind": "plugin", "name": "response-transformer-advanced for route e60ad185-9a86-4ad1-a22f-97c3a93c5db4.multi.echo..1027 and consumer-group test-consumer-group-4"}
```
Looks like the same plugins are created twice.
 I checked with the logs and found that it was passing because of the "two bugs can work" effect:
- The first round of configuration creates the plugins twice and error occurs
- The first round of fallback configuration rollbacks the created plugins
- The second round of configuration still creates the plugins and the error still occurs
- The second round of fallback configuration skips the check because the SHA was not changed compared to the first fallback configuration
So finally the plugins are created and not deleted in the second round.

pmalek commented on internal/dataplane/sendconfig/sendconfig.go:
```diff
@@ -58,8 +58,11 @@ func PerformUpdate(
 
-	// disable optimization if reverse sync is enabled
-	if !config.EnableReverseSync {
+	// Disable the optimization in DB mode by checking if the configuration is same as the last applied configuration if:
+	// - reverse sync is enabled
+	// - or the update is performed by a fallback process because in DBmode, when applying fails, some entities are applied successfully, then the Kong gateway may be in a "partial success" state.
+	if config.InMemory ||
+		(!config.EnableReverseSync && !isFallback) {
```
This changes the current behavior: previously we **only** checked the SHA when reverse sync was disabled (regardless of the db/dbless mode), now we **always** check if in memory mode (because of the **OR** that's used).

I think we can change this to:

```suggestion
	// Disable checking whether the actual SHA is as generated when:
	// - reverse sync is enabled
	// - or running in DB mode and in fallback mode
	//
	// We do this because in db mode, when applying fails, some entities are applied successfully, then the Kong gateway may be in a "partial success" state.
	if !config.EnableReverseSync && (config.InMemory || !isFallback) {
```


Based on:

```
Note: The expression `!EnableReverseSync && (InMemory || !isFallback)` evaluates to 1 if either:
- `EnableReverseSync` is 0 AND `InMemory` is 1, OR
- `EnableReverseSync` is 0 AND `isFallback` is 0
```

```
| EnableReverseSync | InMemory | isFallback | Result |
| ----------------- | -------- | ---------- | ------ |
| 0                 | 0        | 0          | 1      |
| 0                 | 0        | 1          | 0      |
| 0                 | 1        | 0          | 1      |
| 0                 | 1        | 1          | 1      |
| 1                 | 0        | 0          | 0      |
| 1                 | 0        | 1          | 0      |
| 1                 | 1        | 0          | 0      |
| 1                 | 1        | 1          | 0      |
```

randmonkey commented:
Blocked by #7394
